src/Common/src/System/Security/Cryptography/Asn1/GeneralNameAsn.cs

```diff
@@ -0,0 +1,121 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.InteropServices;
+
+namespace System.Security.Cryptography.Asn1
+{
+    // https://tools.ietf.org/html/rfc5280#section-4.2.1.6
+    //
+    // GeneralName ::= CHOICE {
+    //     otherName                 [0]  OtherName,
+    //     rfc822Name                [1]  IA5String,
+    //     dNSName                   [2]  IA5String,
+    //     x400Address               [3]  ORAddress,
+    //     directoryName             [4]  Name,
+    //     ediPartyName              [5]  EDIPartyName,
+    //     uniformResourceIdentifier [6]  IA5String,
+    //     iPAddress                 [7]  OCTET STRING,
+    //     registeredID              [8]  OBJECT IDENTIFIER
+    // }
+    [Choice]
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct GeneralNameAsn
+    {
+        [ExpectedTag(0)]
+        internal OtherNameAsn? OtherName;
+
+        [ExpectedTag(1)]
+        [IA5String]
+        internal string Rfc822Name;
+
+        [ExpectedTag(2)]
+        [IA5String]
+        internal string DnsName;
+
+        [ExpectedTag(3, ExplicitTag = true)]
+        [AnyValue]
+        internal ReadOnlyMemory<byte>? X400Address;
+
+        [ExpectedTag(4, ExplicitTag = true)]
+        [AnyValue]
+        internal ReadOnlyMemory<byte>? DirectoryName;
```

Comment:
This serializes and deserializes properly according to the Pkcs tests and comparisons to [other test data](https://github.com/kjur/jsrsasign/blob/master/test/qunit-do-asn1x509.html). It doesn't match the RFC though and could be confusing. Generally the `ExpectedTag` and `AnyValue` combination is affected by this.

Reply:
The 2012 ITU document (https://www.itu.int/rec/dologin_pub.asp?lang=e&id=T-REC-X.509-201210-S!!PDF-E&type=items) is in `DEFINITIONS IMPLICIT TAGS`.  And I was being massively confused, since both RFC 3280 and 5280 are `DEFINITIONS EXPLICIT TAGS`.

Except, it turns out, that there are TWO modules in those RFCs.  GeneralName is defined in the second one, which is `DEFINITIONS IMPLICIT TAGS`.  Joy.

Reply:
For this one in particular, it stays explicit because of X.680-2015 31.2.7 (c).

> 31.2.7 The tagging construction specifies explicit tagging if any of the following holds:
a) the "Tag EXPLICIT Type" alternative is used;
b) the "Tag Type" alternative is used and the value of "TagDefault" for the module is either EXPLICIT TAGS or is empty;
c) the "Tag Type" alternative is used and the value of "TagDefault" for the module is IMPLICIT TAGS or AUTOMATIC TAGS, but the type defined by "Type" is an untagged choice type, an untagged open type, or an untagged "DummyReference" (see Rec. ITU-T X.683 | ISO/IEC 8824-4, 8.3).

So `field [4] SomeType` where `SomeType ::= CHOICE { ...` is always explicit, because it needs a boundary for the context.

Seems like a thing the serializer can enforce (reject it if ExplicitTag=true isn't declared)

Reply:
I agree it is confusing as hell. I only discovered it when I rewrote `SubjectAlternativeNameBuilder` to use this ASN definition.

Reply:
So I still got `X400Address ORAddress` wrong since it should not have ExplicitTag = true.